### PR TITLE
BFD-2509: RifLoader performance improvements

### PR DESCRIPTION
**Pull request by brianburton**
**JIRA Ticket:**
[BFD-2509](https://jira.cms.gov/browse/BFD-2509)

**User Story or Bug Summary:**

As an operations engineer I want the pipeline to execute weekly CCW RIF loads faster so that I don't have to wait until Sunday to discover that there's a problem with the content of the load.

---

### What Does This PR Do?

- Introduces `DatabaseIdHasher` using combined ram and db cache to avoid recomputing hashes.  Hash computation is extremely slow and CPU intensive and vast majority of IDs are already known to BFD.
- Adds new table, `id_hashes`, as persistent cache.
- Adds `RIF_JOB_BATCH_SIZE` env var to allow experimenting with alternative batch sizes.
- Adds `RIF_JOB_QUEUE_SIZE_MULTIPLE` env var to allow experimenting with less excess executor queue capacity.
- Uses `TransactionManager` to handle commits and retries.
- Adds Micrometer metrics for tracking hash lookups, misses, and retries.
- Removes unused `PostgreSqlCopyInserter` parameter.

**Theory:**

- Setting `RIF_JOB_QUEUE_SIZE_MULTIPLE` lower (e.g. 2 instead of 10) will reduce heap usage substantially without harming performance.
- Setting `RIF_JOB_BATCH_SIZE` lower (e.g. 50 instead of 100) will reduce heap usage without harming performance.

**Note:** First run through with any given data set won't benefit as much from the cache.  Second run should benefit substantially as long as the `id_hashes` table is NOT cleared between runs.

### What Should Reviewers Watch For?

If you're reviewing this PR, please check for these things in particular:

* Verify all PR security questions and checklists have been completed and addressed.

### What Security Implications Does This PR Have?

Submitters should complete the following questionnaire:

* If the answer to any of the questions below is **Yes**, then **you must** supply a link to the associated Security Impact Assessment (SIA), security checklist, or other similar document in Confluence here: **N/A**

    * Does this PR add any new software dependencies? 
      * [ ] Yes
      * [x] No
    * Does this PR modify or invalidate any of our security controls?
      * [ ] Yes
      * [x] No
    * Does this PR store or transmit data that was not stored or transmitted before?
      * [ ] Yes
      * [x] No

* If the answer to any of the questions below is **Yes**, then please add @StewGoin as a reviewer, and note that this PR **should not be merged** unless/until he also approves it.
    * Do you think this PR requires additional review of its security implications for other reasons?
      * [ ] Yes
      * [x] No

### What Needs to Be Merged and Deployed Before this PR?

This PR cannot be either merged or deployed until the following prerequisite changes have been fully deployed:

* N/A

### Submitter Checklist

I have gone through and verified that...:

* [x] I have named this PR and branch so they are [automatically linked](https://confluence.atlassian.com/adminjiracloud/integrating-with-development-tools-776636216.html) to the (most) relevant Jira issue. Ie: `BFD-123: Adds foo`
* [x] This PR is reasonably limited in scope, to help ensure that:
    1. It doesn't unnecessarily tie a bunch of disparate features, fixes, refactorings, etc. together.
    2. There isn't too much of a burden on reviewers.
    3. Any problems it causes have a small "blast radius".
    4. It'll be easier to rollback if that becomes necessary.
* [x] This PR includes any required documentation changes, including `README` updates and changelog / release notes entries.
* [x] The data dictionary has been updated with any field mapping changes, if any were made.
* [x] All new and modified code is appropriately commented, such that the what and why of its design would be reasonably clear to engineers, preferably ones unfamiliar with the project.
* [x] All tech debt and/or shortcomings introduced by this PR are detailed in `TODO` and/or `FIXME` comments, which include a JIRA ticket ID for any items that require urgent attention.
* [x] Reviews are requested from both:
    * At least two other engineers on this project, at least one of whom is a senior engineer or owns the relevant component(s) here.
    * Any relevant engineers on other projects (e.g. DC GEO, BB2, etc.).
* [x] Any deviations from the other policies in the [DASG Engineering Standards](https://github.com/CMSgov/cms-oeda-dasg/blob/master/policies/engineering_standards.md) are specifically called out in this PR, above.
    * Please review the standards every few months to ensure you're familiar with them.

### CODE REVIEW

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/DatabaseIdHasher.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/DatabaseIdHasher.java:25]: It would be trivial to remove and revisit later.

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoader.java
Comment (lsmitchell) [inline apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoader.java:583]: Was this hsql-specific handling removed because the transaction handler handles it for us, or because we've switched to using postgres as our primary testing db? (Asking since hsql is still supported so wondering if this still might need to exist)

**Review by lsmitchell — COMMENTED** [review]
One question, but otherwise looks good; just waiting on final testing to approve.

File: apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoader.java
Comment (brianburton) [reply apps/bfd-pipeline/bfd-pipeline-ccw-rif/src/main/java/gov/cms/bfd/pipeline/ccw/rif/load/RifLoader.java:583]: The `TransactionManager` takes care of handling errors during rollback so this code is obsolete.

**Review by lsmitchell — DISMISSED** [review]
Rollback fix looks ok! I know theres some final testing happening, but approved unless something critical comes from it.

**Review by mjburling — DISMISSED** [review]
The results are sane from this afternoon's sanity checks conducted by @cbrunefearless. I think we're in a good position to accept this. 🙇

**Review by mjburling — APPROVED** [review]

**Review by cbrunefearless — APPROVED** [review]
LGTM

**Pull request merged**
